### PR TITLE
kernel: add MultiPath TCP kernel config & module

### PR DESCRIPTION
**Pull request by sKyissKy**
add MultiPath TCP kernel config & module

### CODE REVIEW

**Comment by Ansuel** [comment]
Hi we support 5.15 and 6.1 kernel version. You can drop each check for the kernel version. Also please drop the config-kernel option and lets just add the kernel modules.

Also please check the guideline on how to submit a patch with adding to this a correct tag and adding a commit description.

File: config/Config-kernel.in [hunk 1]
```diff
@@ -1054,6 +1054,13 @@ if KERNEL_IP_MROUTE
 
 endif
 
+config KERNEL_MPTCP
+	bool "Enable IPv4 MultiPath TCP support"
+	default y
```
Comment (luizluca) [inline config/Config-kernel.in:1059]: default y? How much extra code will it add to the image?

**Comment by john-tho** [comment]
@sKyissKy, you have already been told not to spam pull requests:
https://github.com/openwrt/openwrt/pull/11147#issuecomment-1304643417
https://github.com/openwrt/openwrt/pulls?q=is%3Apr+author%3AsKyissKy+is%3Aclosed
You just waste people's time and compute resources, especially when you are being told exactly the same things: https://github.com/openwrt/openwrt/pull/11147#issuecomment-1453078866. https://github.com/openwrt/openwrt/pull/11064#issuecomment-1288217320
You can use `git rebase` to edit your commits, then `git push --force` to update your PR branch.

If you had addressed the feedback in any of your previous pull requests, this would have been merged a long time ago.

**Comment by sKyissKy** [comment]
> Hi we support 5.15 and 6.1 kernel version. You can drop each check for the kernel version. Also please drop the config-kernel option and lets just add the kernel modules.
> 
> Also please check the guideline on how to submit a patch with adding to this a correct tag and adding a commit description.

If set MPTCP, MPTCP_IPV6 as a kernel module, an empty packet will be generated.

File: config/Config-kernel.in (hunk 1 above)
Comment (sKyissKy) [reply config/Config-kernel.in]: already fixed.

**Comment by john-tho** [comment]
You need to depend on your Config-kernel.in option in your kernel package.
Otherwise, because `MPTCP` is a Kconfig bool, buildbot would compile this into every kernel via `ALL_KMODS`.

This was detailed months ago on one of your duplicate PRs; https://github.com/openwrt/openwrt/pull/11147#issuecomment-1303170026

**Comment by luizluca** [comment]
A PR should not be a log of your changes over time but the logical blocks required to apply your changes. If you need to fix an old commit, you need to squash them together (rebase and squash/fixup). Afterwards, you can push -f your current final work for review. There are tons of instructions on the internet.

The commit titles are not using the correct prefixes used by openwrt. Check the git log and you'll see there is a logic. You can see the tags prefixes by using git log -- aDir/aFile.

You need to describe, at least, what and why in your commit messages. "Update file foo" says nothing. Again, check the existing log to have an idea.

**Comment by sKyissKy** [comment]
> A PR should not be a log of your changes over time but the logical blocks required to apply your changes. If you need to fix an old commit, you need to squash them together (rebase and squash/fixup). Afterwards, you can push -f your current final work for review. There are tons of instructions on the internet.
> 
> The commit titles are not using the correct prefixes used by openwrt. Check the git log and you'll see there is a logic. You can see the tags prefixes by using git log -- aDir/aFile.
> 
> You need to describe, at least, what and why in your commit messages. "Update file foo" says nothing. Again, check the existing log to have an idea.

I edit the code directly on the github webpage. It seems that the github webpage version does not provide the function of rebase and merge commit. If the above functions can be realized on the webpage, please advise.

**Comment by luizluca** [comment]
> > A PR should not be a log of your changes over time but the logical blocks required to apply your changes. If you need to fix an old commit, you need to squash them together (rebase and squash/fixup). Afterwards, you can push -f your current final work for review. There are tons of instructions on the internet.
> > 
> > The commit titles are not using the correct prefixes used by openwrt. Check the git log and you'll see there is a logic. You can see the tags prefixes by using git log -- aDir/aFile.
> > 
> > You need to describe, at least, what and why in your commit messages. "Update file foo" says nothing. Again, check the existing log to have an idea.
> 
> I edit the code directly on the github webpage. It seems that the github webpage version does not provide the function of rebase and merge commit. If the above functions can be realized on the webpage, please advise.

I don't believe it's possible. GitHub web is very basic. You need to edit it locally.

File: package/kernel/linux/modules/netsupport.mk [hunk 2]
```diff
@@ -1434,6 +1434,23 @@ endef
 $(eval $(call KernelPackage,inet-diag))
 
 
+define KernelPackage/inet-mptcp-diag
+  SUBMENU:=$(NETWORK_SUPPORT_MENU)
+  TITLE:=INET diag support for MultiPath TCP
+  DEPENDS:= +@KERNEL_MPTCP +@KERNEL_MPTCP_IPV6 +kmod-inet-diag
+  KCONFIG:= CONFIG_INET_MPTCP_DIAG@ge5.6
+  FILES:= $(LINUX_DIR)/net/mptcp/mptcp_diag.ko@ge5.6
```
Comment (john-tho) [inline package/kernel/linux/modules/netsupport.mk:1442]: no `+` before the `@CONFIG_SYMBOLS`, `+` is for adding packages.

no need for the `@ge5.6` kernel 5.6 version switches after Kconfig symbol nor in file list (this branch now supports 5.15 and 6.1)

**Comment by john-tho** [comment]
Great to see a single commit with an appropriate title now.
You need to add a description to the commit message (explain why this change was made),  and your Signed-off-by tag.

**Comment by Neustradamus** [comment]
To follow

**Comment by Djfe** [comment]
for what @john-tho asked:
commandline:
`git commit --amend --sign-off`
it edits the last commit message and adds your sign off for you :)

**Comment by xback** [comment]
@sKyissKy 
Hi,
Just wondering if you will finish this one? :-)
I'm interested

Thanks!

**Comment by rodsmar** [comment]
I've made a PR based on this. Check #14323

File: config/Config-kernel.in [hunk 3]
```diff
@@ -1134,6 +1134,12 @@ if KERNEL_IP_MROUTE
 
 endif
 
+config KERNEL_MPTCP
+	bool "Enable IPv4 MultiPath TCP support"
+	default n
```
Comment (hauke) [inline config/Config-kernel.in:1139]:
```suggestion
```
`default n` is not needed, `n` is already the default.

File: config/Config-kernel.in [hunk 4]
```diff
@@ -1175,6 +1181,13 @@ if KERNEL_IPV6
 	config KERNEL_LWTUNNEL_BPF
 		def_bool n
 
+	config KERNEL_MPTCP_IPV6
+		bool "Enable IPv6 MultiPath TCP support"
+		default n
```
Comment (hauke) [inline config/Config-kernel.in:1186]:
```suggestion
		default y
```
I assume that most users that have selected `KERNEL_MPTCP` also want to have IPv6 support. 
Please make this active by default.

File: package/kernel/linux/modules/netsupport.mk [hunk 5]
```diff
@@ -1495,6 +1495,23 @@ endef
 $(eval $(call KernelPackage,inet-diag))
 
 
+define KernelPackage/inet-mptcp-diag
+  SUBMENU:=$(NETWORK_SUPPORT_MENU)
+  TITLE:=INET diag support for MultiPath TCP
+  DEPENDS:= +KERNEL_MPTCP +kmod-inet-diag
```
Comment (hauke) [inline package/kernel/linux/modules/netsupport.mk:1501]: Please do not add the `KERNEL_MPTCP` configuration here, bgut depend on it:
```suggestion
  DEPENDS:= @KERNEL_MPTCP +kmod-inet-diag
```

**Comment by hauke** [comment]
Please add your real name as author and in the `Signed-off-by:` line.

**Comment by matttbe** [comment]
Hello,

I guess this PR can be closed now that #16786 has been merged.

**Comment by hauke** [comment]
Multipath TCP support was added in c8d5abd62b70137c70bf12e83b6d0708b980abb3 and a2aabc9a7a4ec8a66477fbdde84a73b5942b460e using #16786.

**Pull request closed without merging**